app/post.go

```diff
@@ -89,7 +97,74 @@ func (a *App) CreatePostMissingChannel(post *model.Post, triggerWebhooks bool) (
 	return a.CreatePost(post, channel, triggerWebhooks)
 }
 
-func (a *App) CreatePost(post *model.Post, channel *model.Channel, triggerWebhooks bool) (*model.Post, *model.AppError) {
+// deduplicateCreatePost attempts to make posting idempotent within a caching window.
+func (a *App) deduplicateCreatePost(post *model.Post) (foundPost *model.Post, err *model.AppError) {
+	// We rely on the client sending the pending post id across "duplicate" requests. If there
+	// isn't one, we can't deduplicate, so allow creation normally.
+	if post.PendingPostId == "" {
+		return nil, nil
+	}
+
+	const unknownPostId = ""
+
+	for attempt := 1; attempt <= 3; attempt++ {
+		// Query the cache atomically for the given pending post id, saving a record if
+		// it hasn't previously been seen.
+		value, loaded := seenPendingPostIds.GetOrAddWithDefaultExpires(post.PendingPostId, unknownPostId)
+
+		// If we were the first thread to save this pending post id into the cache,
+		// proceed with create post normally.
+		if !loaded {
+			return nil, nil
+		}
+
+		postId := value.(string)
+
+		// If another thread saved the cache record, but hasn't yet updated it with the
+		// actual post id, wait a bit and try again. If the cache record expires, we'll
+		// still end up with a duplicate post as expected. Note that we'll only ever be
+		// doing any of this when racing on duplicate create post requests.
+		if postId == unknownPostId {
```

Comment:
Hmm, the mutex would have to be allocated and acquired for every post made, which seems like an unreasonable pressure on memory and performance for what is already an edge case.

Perhaps I could leverage sync.Cond here instead to achieve the same semantics without a sleep. I’ll investigate — thanks!

Reply:
Allocated - yes, but it's just 8 bytes. Acquired - yes, but the first time it's a quick path (https://golang.org/src/sync/mutex.go#L72) and there should be very few after that.  Never used `sync.Cond` before but if it makes things simpler, sure!

Reply:
I've pushed a set of changes that eliminates the sleep, but also eliminates the graceful handling of the window of time between when the cache records the pending post id, and we actually have it saved into the database. Deduplication still occurs, but the client in that window will receive an error for their duplicate post vs. a `201 Created` that returns the original post created.

I couldn't envision a simple way to implement the desired behaviour using `sync`. I wouldn't want just want one `sync.Mutex` here -- otherwise all posting would be serialized -- but really one per pending post id in the cache. So I'd have to allocate it for every incoming post, then store it in the cache (atomically), then have the racing threads in this edge case lock it (blocking until the first thread finishes). This seemed like too much machinery.

I considered `sync.Cond` here and having a model where all threads in this "pending" state are blocked together on a single mutex, with a broadcast to wake them all up when /any/ post is completed. My worry here was the pathological edge case of thousands of posts per second combined with some kind of slowdown that resulted in lots of deduplication, and tons of goroutines waking up over and over as each post made its way through to check if their post had been duplicated or not.

The latest implementation dispenses with all of that by essentially leaving it up to the client to handle the error in this window, if desired. Without special handling, there would be an error message interaction, but we'd still have the guarantee of deduplication. This seems like the best compromise.

Open to further discussion and ideas here, as I might be missing a simple solution :)